source/core/DualTypeConversion.cpp

```diff
@@ -0,0 +1,316 @@
+    bool DualTypeConversion::ApplyS2CInts(TypeRef typeRefX, void* pDataX, TypeRef typeRefY, void* pDataY)
+    {
+        switch (typeRefX->TopAQSize()) {
+            case 0:
+                break;
+            case 1: {
+                Int8 int8ValueX = *static_cast<Int8*>(pDataX);
+                ApplyNumeric<Int8>(int8ValueX, typeRefY, pDataY);
+                break;
+            }
+            case 2: {
+                Int16 int16ValueX = *static_cast<Int16*>(pDataX);
+                ApplyNumeric<Int16>(int16ValueX, typeRefY, pDataY);
+                break;
+            }
+            case 4: {
+                Int32 int32ValueX = *static_cast<Int32*>(pDataX);
+                ApplyNumeric<Int32>(int32ValueX, typeRefY, pDataY);
+                break;
+            }
+            case 8: {
+                Int64 int64ValueX = *static_cast<Int64*>(pDataX);
+                ApplyNumeric<Int64>(int64ValueX, typeRefY, pDataY);
+                break;
+            }
+        }
+        return true;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::ApplyIEEE754Binaries(TypeRef typeRefX, void* pDataX, TypeRef typeRefY, void* pDataY)
+    {
+        if (typeRefX->TopAQSize() == sizeof(Single)) {
+            Single singleValueX = *reinterpret_cast<Single*>(pDataX);
+            ApplyNumeric<Single>(singleValueX, typeRefY, pDataY);
+        } else {
+            Double doubleValueX = *reinterpret_cast<Double*>(pDataX);
+            ApplyNumeric<Double>(doubleValueX, typeRefY, pDataY);
+        }
+        return true;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::AreBooleanCompatible(TypeRef typeRefX, TypeRef typeRefY)
+    {
+        bool sameEncodingAndSize = DoTypesHaveSameEncodingAndSize(typeRefX, typeRefY);
+        return sameEncodingAndSize;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::AreUIntCompatible(TypeRef typeRefX, TypeRef typeRefY)
+    {
+        bool typesAreCompatible = TypesAreCompatible(typeRefX, typeRefY);
+        return typesAreCompatible;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::AreS2CIntCompatible(TypeRef typeRefX, TypeRef typeRefY)
+    {
+        bool typesAreCompatible = TypesAreCompatible(typeRefX, typeRefY);
+        return typesAreCompatible;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::AreIEEE754BinaryCompatible(TypeRef typeRefX, TypeRef typeRefY)
+    {
+        bool typesAreCompatible = TypesAreCompatible(typeRefX, typeRefY);
+        return typesAreCompatible;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::AreIntrinsicClustersCompatible(TypeRef typeRefX, TypeRef typeRefY)
+    {
+        if (typeRefX->IsComplex() && typeRefY->IsComplex()) {
```

Comment:
We need to check for only Complex here? How about Timestamp type? Could you explain/add a comment why this is necessary?

Reply:
Because `IsComplex` covers both `ComplexSingle` and `ComplexDouble`, which would fail the name check below although they're actually compatible for conversion purposes. It is OK for timestamp to go through the name check below.

Added comment.